add transformer and checkedTransformer

phlptp opened the pull request:
This is the last of the PR's originating from #229.  
This builds on the IsMember function and adds Transformer and CheckedTransformer validators  

Modifications include disabling the IsMember validator from using the second element of a map, so maps with non-convertible types can be used.  IsMember will modify the input to match an element of the set.

Validators added through the transform function are inserted at the beginning of the validators list, this allows transformations to occur before a check regardless of the order added.  

Some tweaks and updates to the IsMember function to clean up some loops and remove some unnecessary processing from inside the loop.

phlptp committed:
add transform and checkedTransform tests

add Transformer and CheckedTransformer validators

henryiii commented on tests/TransformTest.cpp:
```diff
@@ -0,0 +1,245 @@
+    EXPECT_EQ(value, test::val3);
+
+    args = {"-s", "val4"};
+    EXPECT_THROW(run(), CLI::ConversionError);
```
Add `args = {"-s", "5"};` here too

phlptp replied:
That won't produce any errors, since Transformer explicitly doesn't do any checking, and 5 is still a valid member of the underlying type.  Is the proposed check just to verify that?

henryiii replied:
Yes. Tests for things that pass are good too, plus it's an example of what's different.

phlptp committed:
Eliminate the Validator description string, some code cleanup

add tests

Make Validators a full Object and remove friend,  move to descriptions instead of overriding type name.

update validators to actually merge the type strings and use all validators in the type outputs

rework join so it works without the start variable,  allow some forwarding references in the validator types, some tests for non-copyable maps, and transforms

merge the search function and enable use of member search function,  make the pair adapters forwarding instead of copying

henryiii commented on include/CLI/Validators.hpp:
```diff
@@ -32,37 +32,75 @@ class Option;
   public:
+    Validator() = default;
+    // Construct Validator from basic information
+    explicit Validator(std::function<std::string(std::string &)> op, std::string name = "", std::string desc = "") {
```
```suggestion
    explicit Validator(std::function<std::string(std::string &)> op, std::string name = "", std::string desc = "")
        : func(std::move(op)), name_(std::move(name)), description_(std::move(desc)): {}
```

(Note: needs to be formatted correctly) (this will make Codacity happy)

phlptp replied:
fixed

henryiii commented on include/CLI/Validators.hpp:
```diff
@@ -32,37 +32,75 @@ class Option;
-    /// This is the type function, if empty the tname will be used
-    std::function<std::string()> tname_function;
+    /// This is the description function, if empty the description_ will be used
+    std::function<std::string()> desc_function;
```
Please use the convention that member variables end in underscore for `desc_function` and `func`.

henryiii commented on include/CLI/Validators.hpp:
```diff
@@ -32,37 +32,75 @@ class Option;
-    /// This is the type name, if empty the type name will not be changed
-    std::string tname;
+    /// This is the validator description
+    std::string description_;
```
Could this be removed and all descriptions come from a function? (At one point I did different things based on whether it was a function or not, but I'm not sure you do)

phlptp committed:
add a few more tests and documentation

fix some gcc 4.7 issues and add a few more test cases and more parts of the README

Work on ReadMe and add Bound validator to clamp values

henryiii commented on README.md:
```diff
@@ -334,7 +316,110 @@ You can access a vector of pointers to the parsed options in the original order
 everything after that is positional only.
 
-#### Getting results 🚧
+#### Validators
```
TOC needs change too?

phlptp replied:
I think the Readme needs another pass especially with the option_group merge, so I can double check to make sure the TOC matches the headers,  Would you want all the subtitles in the TOC?

henryiii replied:
Unless it makes it too long, I think it would be fine.

henryiii commented on README.md:
```diff
@@ -334,7 +316,110 @@ You can access a vector of pointers to the parsed options in the original order
+
+-   `CLI::IsMember(...)`: 🚧 Require an option be a member of a given set. See below for options.
+-   `CLI::Transformer(...)`: 🚧 Modify the input using a map. See below for options.
+-   `CLI::CheckedTransformer(...)`: 🚧 Modify the input using a map, and verify that the input is either in the set or already one of the outputs of the set. See below for options.
```
What about the `->unchecked` idea?

phlptp replied:
There is an active(false) function on validators now.  That is the same concept,  It could be renamed as checked(), unchecked(), is_checked(), if you think that would read better and be more intuitive.

phlptp committed:
updates to README.md

phlptp committed:
Add some more in TOC of README and fix style in Option.hpp
